Version of Random.choice() that returns indices

Aniket21mathur opened the pull request:
This pr:

- Updates the `_choice` algorithm to support a domain instead of an array.
- Add domain and range overloads.
- Modify array overload to pass domain to `_choice`.  
- Fix spell typo in `proc` descriptions. 

Fixes #14865

Aniket21mathur commented:
@bradcray @ben-albrecht I have made the changes as we discussed in the issue thread. I did basic checking for both arrays and ranges with 

```
use Random;

var rng = new owned RandomStream( eltType= real, seed= 200 );
//var rng = new owned RandomStream( eltType= int, seed= 200 );   // this also works but not sure about details...

var arr = ["a", "b", "c"];
var r = 1..3;
var n = 20;

writeln();
for i in 1..3 do writeln( rng.choice( r ));

writeln();
for i in 1..3 do writeln( rng.choice( r, size= 3 ) );

writeln();
for i in 1..3 do writeln( rng.choice( r, size= 3, replace= false ) );

```

I will update/add tests once the changes are approved. :smile: 

Please review.

Thanks!

bradcray commented:
Hi @Aniket21mathur: Thanks for the update.  I don't know about @ben-albrecht, but I often approach a review by first looking at the tests that are being filed with the PR to see examples of the feature in use and a case I can look at and try as I look through the code (with bonus points for having the test(s) demonstrate corner cases and error conditions).

Aniket21mathur commented on modules/standard/Random.chpl:
```diff
@@ -324,20 +328,21 @@ module Random {
-    if arr.domain != prob.domain then
-      throw new owned IllegalArgumentError('choice() arrays must have equal domains');
+    /*
+    if X != prob.domain then
```
How should we handle this, as the original array is reindexed in the overload function itself?

ben-albrecht replied:
I think it's OK for stride or offset to not match for the domain/range case.

Aniket21mathur replied:
thanks. I have removed the check

Aniket21mathur commented on modules/standard/Random.chpl:
```diff
@@ -250,44 +250,47 @@ module Random {
 
-      if prob.domain != arr.domain {
-        throw new owned IllegalArgumentError('choice() array arguments must have same domain');
+      //Commenting out for now, as X is domain of reindexed array.
```
How should we handle this, as the original array is reindexed in the overload function itself?

ben-albrecht replied:
see other answer

Aniket21mathur commented:
@bradcray @ben-albrecht 

The progress is as follows:-

**For Arrays:-**
- Working fine with strided case.
- Working fine with probability array.
- One thing left is comparing the domain of the input array and the probability array, I have left a comment as a query, please see.

All tests in `/test/library/standard/Random/RandomStreamInterface/choiceTest.chpl` pass.

**For Ranges:-**
- For ranges, we have to explicitly define whether a range is strided or not, so do we have to define 2 seperate overloads for the two cases?
- For handling strided ranges, I am converting a range into an array and then passing it to the array overload. Please see the single line comment.
- For ranges like 0..3, the domain.alignedLow value will be 0 for probability array it will be something > 0, in that case the domain values of the two can never be equal. What should we do in this case?

Thanks :smile:

Aniket21mathur commented on modules/standard/Random.chpl:
```diff
@@ -986,9 +991,50 @@ module Random {
+      proc choice(rng: range(stridable=true), size:?sizeType=none, replace=true, prob:?probType=none)
+        throws
+      {
+        var A: domain(1,stridable=true) = {rng};
```
I have commented this overload for now, it is for strided ranges. The ranges are converted into array and then it is passed to the array overload.

ben-albrecht replied:
What is the intention with this overload? 

> The ranges are converted into array and then it is passed to the array overload.

Note we want the implementation that _doesn't_ create a temporary array.

Aniket21mathur replied:
> What is the intention with this overload?

@ben-albrecht as you suggested to add an overload for handling ranges as well, I have added this overload. As we can also have strided ranges just like  strided arrays, to handle those cases and in general I have converted the range into an array and passed it to the array overload.

Aniket21mathur replied:
> Note we want the implementation that _doesn't_ create a temporary array.

We are handling arrays in the overloads only. Operations in the `_choice` function are peformed on a domain only. Since we have to take care of the strided cases and we are returning indices now we have to deal with the `low` value and the `stride` of an array/range.

ben-albrecht commented on test/library/standard/Random/RandomStreamInterface/choiceTest.chpl:
```diff
@@ -84,7 +84,7 @@ record R {
 
 proc testArray(stream, arr: [] ?eltType, size:?sizeType=none, replace=true, prob:?probType=none, trials=10000) throws {
-  var counts = new map(eltType, int);
+  var counts = new map(int, int);
```
I think this is reverting a change made on master.

Aniket21mathur replied:
This seems to be necessary, as now we have to maintain a key -> value pair of (int, int) instead of (element type, int).
(Considering the fact that we have to check for `indices` rather than `elements` now.)

ben-albrecht replied:
huh, strange. Can you pull or rebase with master to confirm the latest changes to `Random.choice()` work with this change?

Aniket21mathur replied:
I rebased with master(after pulling from the upstream master), and all tests in ` test/library/standard/Random/RandomStreamInterface/choiceTest.chpl` pass. 

I am sorry, but I am getting a bit confused, why is this change so surprising? Currently, in `choice` we are returning elements of the array, which can be `int` or `char`, so we used `eltType` of the array. But in this version of `choice` we are returning indices, that should be `int` so we used `int` type for the map key, which otherwise will throw an error.

Aniket21mathur commented:
@ben-albrecht Thanks for reviewing. I have made some changes and posted some inline responses please have a look.

ben-albrecht commented:
> For ranges we have to explicitly define a range to be stridable, if a non stridable range is passed an exception is thrown. In this case what should we do? A user might give a strided or a non-strided
range as an input.

We _should_ be able to handle strided ranges as well, but we will have to handle the stride math in the implementation, which contributes to making this more difficult to implement 😈 

I believe the `Sort` and maybe some other `Random` functions have some examples of code dealing with strided ranges/domains if that's helpful.

ben-albrecht commented:
You will also want to confirm that the range is [bounded](https://chapel-lang.org/docs/master/builtins/ChapelRange.html#ChapelRange.BoundedRangeType). An unbounded range should throw an error.

Aniket21mathur commented:
@ben-albrecht 

> You will also want to confirm that the range is [bounded](https://chapel-lang.org/docs/master/builtins/ChapelRange.html#ChapelRange.BoundedRangeType). An unbounded range should throw an error.

Thanks, handled the case in the range overload.


> We _should_ be able to handle strided ranges as well. We may be able to avoid doing the stride math by utilizing [orderToIndex](https://chapel-lang.org/docs/master/builtins/ChapelRange.html#ChapelRange.range.orderToIndex).
> 
> I believe the `Sort` and maybe some other `Random` functions have some examples of code dealing with strided ranges/domains if that's helpful.

Yes, made modifications accordingly to support both strided arrays and strided ranges without using 
a reindexed array. Please review.

Before adding tests for ranges I have a doubt. The current range overload takes an strided range as an input from the user,

```chapel
proc choice(rng: range(stridable=true), size:?sizeType=none,
```
Use case

```chapel
var rng = new owned RandomStream( eltType= real, seed= 200 );
var rD : range(stridable=true) = 0..10 by 2;
writeln(rng.choice(rD));
```
But if the user gives a non stridable range as an input, it throws an unresolved call error.

```chapel
var rng = new owned RandomStream( eltType= real, seed= 200 );
var rD : range = 0..10;
writeln(rng.choice(rD));
```
So should we add a separate overload for the non stridable ranges?

ben-albrecht commented:
@Aniket21mathur - you can add a new overload, or, if it would be cleaner, just make the current overload generic w.r.t. stridable:

```
proc choice(rng: range(stridable=?), size:?sizeType=none, ...)
```

Then check for stride where necessary in the function.

Aniket21mathur commented:
@ben-albrecht @bradcray I have added tests for range as well. I have renamed tests for array from `choiceTest` to `choiceTestArray` and made `choiceTestRange` files.

I request you to review all the changes and give your suggestions.

Thanks :-)

cassella commented on modules/standard/Random.chpl:
```diff
@@ -973,21 +977,56 @@ module Random {
         throws
       {
-        return _choice(this, arr, size=size, replace=replace, prob=prob);
+        return _choice(this, Dom, size=size, replace=replace, prob=prob);
```
`_choice(Dom,...)` returns an index or indices from the `Dom`, right?  This function `choice(arr: [?Dom],...)` should return elements of the array, not indices of its domain.

Aniket21mathur replied:
@cassella thanks for reviewing. The `_choice(Dom, ...)` function do not return indices, it returns elements from `Dom` which are indices of the passed array/range.

cassella replied:
Right.  But the `choice(arr: [?Dom],)` function should return elements of `arr`, not elements of `Dom`.

What does `stream.choice(["a"])` return?

Aniket21mathur replied:
> What does `stream.choice(["a"])` return?

@cassella this will return `0`.

> But the `choice(arr: [?Dom],)` function should return elements of `arr`, not elements of `Dom`.

I think this function was expected to return indices of `arr` in the new implementation, which are elements of `Dom` of that array. In the `_choice()` function we are randomly generating an index for the `Dom` and then mapping it to an element of the `Dom` which in turn is an indice of the `arr`.

cassella replied:
The existing function `choice(arr: [?Dom],...)` returns an array element.

I think the request in #14865 is to add two new versions of
`choice()`: one taking a domain and another taking a range, that
return indices.  I believe it was requesting those in addition to the
existing behavior.

That is, ending up with three versions:

`choice(arr: [?Dom], ...)` - returns an array element
`choice(D: domain, ...)` - returns an index from the domain
`choice(r: range, ...)` - returns an index from the range.

And so

```chapel
var A : [D] string;
stream.choice(A); // returns a string
stream.choice(A.domain); // returns an index in D
```

I don't see anything in the request suggesting to change the behavior
of existing calls to `choice(arr[])`.


I think this is why @ben-albrecht was expecting that `var counts = new
map(eltType, int)` in the array test case to continue counting `eltType`s.

Aniket21mathur replied:
@cassella thanks for explaining, I think I misinterpreted @ben-albrecht's [comment](https://github.com/chapel-lang/chapel/issues/14865#issuecomment-585764974). I thought we do not want to keep the old functionality, as random elements of an array can be obtained if we have random indices. Currently, we have two overloads some for `arr` and one for `rng`,
and `_choice` accepts domain and return random elements from the domain.

Now to return elements in case of an array, the array should also be passed to the `_choice` function. But in case of domain and range where we have to return indices there is no need of an array. Should I add an optional array arg to `_choice` just like the `prob` arg? What do you suggest?

cassella replied:
That's just my take; you can wait for Ben or Brad to clarify.

I don't think you'd have to make that drastic a change -- you're almost there already:

```chapel
proc choice(arr: [?Dom], ...) {
    var idx = _choice(this, Dom, ...);
    return arr[idx];
}
```

Aniket21mathur replied:
> ```chapel
> proc choice(arr: [?Dom], ...) {
>     var idx = _choice(this, Dom, ...);
>     return arr[idx];
> }
> ```

This seems to be a good approach. Thanks!

> That's just my take; you can wait for Ben or Brad to clarify.

Yes, let's ask @ben-albrecht and @bradcray what they think?

ben-albrecht replied:
@cassella's interpretation is correct:

That is, ending up with three versions:

> `choice(arr: [?Dom], ...)` - returns an array element
> `choice(D: domain, ...)` - returns an index from the domain
> `choice(r: range, ...)` - returns an index from the range.

Aniket21mathur replied:
> That is, we should end up with three overloads:

I will make the necessary changes. Thanks :-)

ben-albrecht replied:
OK, feel free to ping me when ready for review again.

Aniket21mathur replied:
> OK, feel free to ping me when ready for review again.

@ben-albrecht I have made the changes. Added overload for domain and modified array overload to return elements like before. I added/updated the tests accordingly ;-)

cassella commented on modules/standard/Random.chpl:
```diff
@@ -973,21 +977,56 @@ module Random {
               domain.
 
      :throws IllegalArgumentError: if ``arr.size == 0``,
                                    if ``arr`` contains a negative value,
```
Here's a pre-existing problem with the doc comment -- `arr` can have negative values, or be all 0.  It's `prob` that has these restrictions.

Aniket21mathur replied:
Yeah! Thanks for pointing out. I will fix that :-)

ben-albrecht commented on test/library/standard/Random/RandomStreamInterface/choiceTestRange.chpl:
```diff
@@ -0,0 +1,131 @@
+
+
+  // offset & strided range
+  var strideRng = 2..10 by 2;
```
Can you integrate these test cases into the existing test for `choice`?

ben-albrecht commented on test/library/standard/Random/RandomStreamInterface/choiceTestDomain.chpl:
```diff
@@ -0,0 +1,131 @@
+use Random, Map;
```
Can you integrate these test cases into the existing test for choice?

Aniket21mathur replied:
> Can you integrate these test cases into the existing test for choice?

@ben-albrecht I thought it would be cleaner to keep the tests in separate files :sweat_smile: .

Though if we keep the tests in the same file, the `testx` (where x can be domain, range o array will require changes, it will involve some conditional statements depending on the argument being an array, domain or range. Some print statements will also need to be standardized without  explicitly mentioning the `arg` type.

ben-albrecht replied:
Well, the test programs have a lot of common functionality that gets duplicated by splitting them up. By merging them and adding the conditional statements needed, we can improve the maintainability of the tests.

We could also keep the tests in separate files and place the common code into a non-tested file like `choiceTestUtils.chpl` with a  `choiceTestUtils.notest`.

Aniket21mathur replied:
> Well, the test programs have a lot of common functionality that gets duplicated by splitting them up. By merging them and adding the conditional statements needed, we can improve the maintainability of the tests.

Yes! agreed. 

> We could also keep the tests in separate files and place the common code into a non-tested file like `choiceTestUtils.chpl` with a `choiceTestUtils.notest`.

Yeah, this is a good alternative :smile:

Aniket21mathur replied:
@ben-albrecht the `test` function code that is common between  `range`, `array` and `domain` do a comparison between the `actualRatios` and the `expectedRatios` obtained from the `prob` array. There is no functionality that requires specific elements of the `array` other than a couple of print statements, which I think can be replaced by the corresponding domain elements.

```chapel
  if debug {
    writeln('Counts for array: ', arr);
    for value in counts {
      writeln(value, ' : ', counts[value]/trials:real);
    }
  }
```
and 

```
    write('[', arr, '], ');
```
So, I was thinking of putting the `test` proc in a utils file as you suggested and modifying it to expect only a `domain` as an argument for all three cases.

What do you think?

Thanks!

ben-albrecht replied:
@Aniket21mathur - Yes, that sounds like a good plan to me. Thanks for confirming.

Aniket21mathur replied:
@ben-albrecht I am having some second thoughts on it, if we go the way I described, the `range` and `array` overloads will not be utilized in tests. Also, the tests currently in `master` just `Confirm that resulting ratios are within 0.05 of expected ratios`, keeping that in mind, do we need seperate tests for `range` , `array` and `domain`? Ultimately we are passing a domain to the `_choice` function. Though the error tests will be different for sure.

ben-albrecht replied:
Good point.

Is it possible to make `testArray` generic so that it will accept any range, domain, or array and pass it off to `choice` without doing anything specific to the type?

If necessary, you could split the validation code out into separate overloaded functions to validate the results for each type (it's not clear to me if that's necessary or not though - at a glance).

ben-albrecht commented on modules/standard/Random.chpl:
```diff
@@ -971,23 +975,97 @@ module Random {
+        if !isBoundedRange(rng) then
+          throw new owned IllegalArgumentError('input range must be bounded');
+
+        var Dom: domain(1,stridable=true) = if isBoundedRange(rng) then {rng} else {1..1};
```
Some cleaner alternatives:


```suggestion
        var Dom: domain(1,stridable=true) = if isBoundedRange(rng) then {rng} else none;
```

or

```chpl
var Dom: domain(1, stridable=true);

if !isBoundedRange(rng) then
  throw new owned IllegalArgumentError('input range must be bounded');
else 
  Dom = {rng};
```

ben-albrecht replied:
nit: `Dom` should be lowercase `dom`

ben-albrecht commented on modules/standard/Random.chpl:
```diff
@@ -971,23 +975,97 @@ module Random {
                                    if ``replace=false`` and ``size > arr.size || size.size > arr.size``
      */
-      proc choice(arr: [], size:?sizeType=none, replace=true, prob:?probType=none)
+      proc choice(arr: [?Dom], size:?sizeType=none, replace=true, prob:?probType=none)
```
nit: `Dom` should be `dom`

```suggestion
      proc choice(arr: [?dom], size:?sizeType=none, replace=true, prob:?probType=none)
```

ben-albrecht commented on modules/standard/Random.chpl:
```diff
@@ -301,18 +303,19 @@ module Random {
         }
       } else {
-        var indices: [A.domain] int = A.domain;
+        var indices: [X] int = X;
```
Here is a temporary array we want to avoid creating, because `X` can potentially be very large.

ben-albrecht commented on modules/standard/Random.chpl:
```diff
@@ -359,7 +363,7 @@ module Random {
       var randNum = stream.getNext(resultType=real);
       var (found, idx) = Search.binarySearch(cumulativeArr, randNum);
-      return A[idx];
+      return (idx-1) * stride + low;
```
I wonder if you could use [`range.orderToIndex`](https://chapel-lang.org/docs/master/builtins/ChapelRange.html#ChapelRange.range.orderToIndex) here?

Aniket21mathur replied:
@ben-albrecht before that I have a doubt. After shifting to 0 based indexing can `idx` be zero now? I am not much sure about that.

Aniket21mathur replied:
> I wonder if you could use [`range.orderToIndex`](https://chapel-lang.org/docs/master/builtins/ChapelRange.html#ChapelRange.range.orderToIndex) here?

Yes, depending on start value of `idx` we can do something like 

```chapel
X.dim(0).orderToIndex(idxl);
```
or 

```chapel
X.dim(0).orderToIndex(randVal);
```

ben-albrecht commented on modules/standard/Random.chpl:
```diff
@@ -399,7 +403,7 @@ module Random {
             if !indicesChosen.contains(indexChosen) {
               indicesChosen += indexChosen;
-              samples[i] = A[indexChosen];
+              samples[i] = (indexChosen-1) * stride + low;
```
I wonder if you could use [`range.orderToIndex`](https://chapel-lang.org/docs/master/builtins/ChapelRange.html#ChapelRange.range.orderToIndex) here?

Aniket21mathur replied:
Yes, same goes for here as well :-)

Aniket21mathur commented on modules/standard/Random.chpl:
```diff
@@ -971,23 +975,97 @@ module Random {
+                                   if ``replace=false`` and ``size > rng.size || size.size > rng.size``.
+                                   if ``isBoundedRange(rng) == false``
+     */
+      proc choice(rng: range(stridable=?), size:?sizeType=none, replace=true, prob:?probType=none)
```
> How about we make all the argument names across `choice()` overloads the same, e.g. `x`?

Yes, that will be good I will make the changes :-)

ben-albrecht commented:
> you suggested to prevent _choiceUniform from creating an O(n) temp array. In one way or the other the array data structure here is used to store the already used indices so that we can avoid repetition. For this purpose, I think ultimately we will need some structure with the worst-case time complexity of O(N). What do you suggest?

@Aniket21mathur - I tried to summarize this design decision for discussion over here: https://github.com/chapel-lang/chapel/issues/14865#issuecomment-615372671

Aniket21mathur commented:
@ben-albrecht I have made the suggested changes.

I also did a test to confirm the randomness of the algorithm implemented in 8d5cd7b. I did it with -

```
use Random;
var rng = new owned RandomStream( eltType= real);
var r = {1..50};
for i in 1..3 do writeln( rng.choice( r, size= 3, replace= false ) );

```chapel
Getting a good set of random results.

I have also updated the tests accordingly, making the `test` function more generic.

Please review. Thanks!

ben-albrecht commented on modules/standard/Random.chpl:
```diff
@@ -230,15 +232,15 @@ module Random {
-    if arr.rank != 1 {
-      compilerError('choice() array must be 1 dimensional');
+    if X.rank != 1 {
+      compilerError('choice() domain must be 1 dimensional');
```
Should these error messages be more generic, since `x` can be an array, domain, or range?

Maybe just use identifier `x`?

```suggestion
      compilerError('choice() argument x must be 1 dimensional');
```

Aniket21mathur replied:
Yes, Agreed.

ben-albrecht commented on modules/standard/Random.chpl:
```diff
@@ -230,15 +232,15 @@ module Random {
-    if arr.size < 1 {
-      throw new owned IllegalArgumentError('choice() array.size must be greater than 0');
+    if X.size < 1 {
+      throw new owned IllegalArgumentError('choice() domain.size must be greater than 0');
```
```suggestion
      throw new owned IllegalArgumentError('choice() x.size must be greater than 0');
```

ben-albrecht commented on modules/standard/Random.chpl:
```diff
@@ -250,45 +252,47 @@ module Random {
-      if prob.domain != arr.domain {
-        throw new owned IllegalArgumentError('choice() array arguments must have same domain');
+      if prob.size != X.size {
+        throw new owned IllegalArgumentError('choice() domain.size must be equal to prob.size');
```
```suggestion
        throw new owned IllegalArgumentError('choice() x.size must be equal to prob.size');
```

ben-albrecht replied:
etc. etc. I'll let you fix the other cases :)

ben-albrecht commented on modules/standard/Random.chpl:
```diff
@@ -301,18 +305,33 @@ module Random {
+              var randIdx = X.dim(0).orderToIndex(randVal);
+              samples[i] = randIdx;
+              indices.add(randVal);
+              i = i + 1;
```
```suggestion
              i += 1;
```

ben-albrecht commented on modules/standard/Random.chpl:
```diff
@@ -301,18 +305,33 @@ module Random {
+          var indices: [X] int = X;
+          shuffle(indices);
+          for i in samples.domain {
+            samples[i] = (indices[(i-1)*stride+low]);
```
Could we `orderToIndex` here instead of computing the index manually?

Aniket21mathur replied:
Yes, thanks for pointing out I missed that.

ben-albrecht commented on modules/standard/Random.chpl:
```diff
@@ -325,20 +344,21 @@ module Random {
+    
     if prob.size == 0 then
-      throw new owned IllegalArgumentError('choice() arrays cannot be empty');
+      throw new owned IllegalArgumentError('choice() array cannot be empty');
```
Let's stick with arg names:

```suggestion
      throw new owned IllegalArgumentError('choice() prob cannot be empty');
```

Aniket21mathur replied:
@ben-albrecht wouldn't it be better to use `prob array` as used in other error calls?

ben-albrecht replied:
Yes, good point.

ben-albrecht commented on modules/standard/Random.chpl:
```diff
@@ -325,20 +344,21 @@ module Random {
-
+    
+    if prob.size != X.size {
+      throw new owned IllegalArgumentError('choice() domain.size must be equal to prob.size');
```
```suggestion
      throw new owned IllegalArgumentError('choice() x.size must be equal to prob.size');
```

ben-albrecht commented:
> Getting a good set of random results.

Great! I did a quick randomness test myself, with the `set` version and the `tmp array` mode. The  results looked good to me as well.

```chpl
use Random;
import Map.map;
import List.list;
import Sort.sort;

config const N = 100000;

var arr: [0..<6] int = 0..<6;
var rs = Random.createRandomStream(int);


var hist = new map(string, int);

for i in 0..<N {
  var result = rs.choice(arr, size=3, replace=false);
  sort(result);
  var key = ''.join(result:string);
  hist[key] += 1;
}

for (k, v) in hist.items() {
  writeln(k, ':', v);
}
```

<details>

For reference, the temp array version took 68 seconds and the set version took 49 seconds (we default to the temp array version for these args, which makes me think the log2 comparison can be tuned further in the future).


```
sample,temp,set
0 1 2,499433,498803
0 1 3,500536,501043
0 1 4,500056,499626
0 1 5,498906,499189
0 2 3,500727,501052
0 2 4,499264,500798
0 2 5,500062,499793
0 3 4,500338,499734
0 3 5,499477,499654
0 4 5,500056,499125
1 2 3,498732,500044
1 2 4,499160,501438
1 2 5,501361,500091
1 3 4,499483,499343
1 3 5,500746,500730
1 4 5,499865,500372
2 3 4,499670,499845
2 3 5,500123,499032
2 4 5,501178,500105
3 4 5,500827,500183
```

</details>

Aniket21mathur commented:
@ben-albrecht thanks for reviewing! I made the suggested changes :-)

> I did a quick randomness test myself, with the `set` version and the `tmp array` mode. The results looked good to me as well.

That's great !

> to the temp array version for these args, which makes me think the log2 comparison can be tuned

Yeah! Agreed.

Aniket21mathur commented:
@ben-albrecht the make check is failing with

```
$CHPL_HOME/modules/standard/Random.chpl:69: error: Cannot find symbol 'Set'
Makefile:204: recipe for target 'check' failed
make: *** [check] Error 1
```

ben-albrecht commented:
@Aniket21mathur - can you do a `git pull` or `git rebase master` to get the 1.22 changes into this branch?

Aniket21mathur commented:
> @Aniket21mathur - can you do a `git pull` or `git rebase master` to get the 1.22 changes into this branch?

@ben-albrecht still failing on my local machine with the same error message :confused:

ben-albrecht commented:
>  still failing on my local machine with the same error message

Just sanity checking - did you rebuild after pulling from master?

Aniket21mathur commented:
> Just sanity checking - did you rebuild after pulling from master?

Yeah! I did

ben-albrecht commented:
Just tried it on my local checkout and see the same issue. Strange... I'll look into it. Just stick with `use Set;` for now.

Aniket21mathur commented:
> Just tried it on my local checkout and see the same issue. Strange... I'll look into it. Just stick with `use Set;` for now.

Yeah, I reverted it. The checks pass now.

Aniket21mathur commented:
> can you update the virtual methods on `RandomStreamInterface` to:

@ben-albrecht sure I will, also we also have a method for `NPBRandomStream`, shouldn't we also update that?



> 1. Include the doc strings as well

Sorry, but I am not able to clearly understand this. Are you pointing to update the documentation for the `proc`? In that case, how should we generalize the documentation? 

Thanks!

ben-albrecht commented:
> sure I will, also we also have a method for NPBRandomStream, shouldn't we also update that?

Good point. `NPBRandomStream` doesn't support `choice()` but you could add the range/domain overloads to give the clean error message similar to array.

> Sorry, but I am not able to clearly understand this. Are you pointing to update the documentation for the proc? In that case, how should we generalize the documentation?

Currently, your changes update [`PCGRandomStream.choice()`](https://chapel-lang.org/docs/master/modules/standard/Random/PCGRandom.html#PCGRandom.PCGRandomStream.choice) docs, but not [`RandomStreamInterface.choice()`](https://chapel-lang.org/docs/master/modules/standard/Random.html#Random.RandomStreamInterface.choice) docs. 

I am requesting that you update the existing `RandomStreamInterface.choice()` method to use `x` instead of `arr`, and to add 2 more `RandomStreamInterface.choice()` methods for range and domain with the docs copied from the `PCGRandomStream.choice()` methods.

Aniket21mathur commented:
> I am requesting that you update the existing `RandomStreamInterface.choice()` method to use `x` instead of `arr`, and to add 2 more `RandomStreamInterface.choice()` methods for range and domain

If we are adding two other overloads then is there a need to use `x` in place of `arr`?

ben-albrecht commented:
> If we are adding two other overloads then is there a need to use x in place of arr?

I think so, just to keep the argument names consistent across all the overloads.

Aniket21mathur commented:
> I think so, just to keep the argument names consistent across all the overloads.

Yeah, sorry just slipped from my mind :sweat_smile:

Aniket21mathur commented:
> Looks good! Confirming docs render well and testing once more before approving/merging.

Thanks :-)

ben-albrecht commented on modules/standard/Random.chpl:
```diff
@@ -535,11 +555,77 @@ module Random {
+    /*
+     Returns a random sample from a given 1-D range, ``x``.
+
+     :arg x: a 1-D range with values that will be sampled from.
```
One little nitpick I caught while looking at rendered docs: a range can only be 1 dimensional in Chapel (domains introduce dimensionality by combining multiple ranges). Maybe we should specify bounded range too, so I'd suggest:

`s/1-D range/bounded range/g` for all the `choice()` docs, e.g.

```suggestion
     :arg x: a bounded range with values that will be sampled from.
```

Aniket21mathur replied:
> I caught while looking at rendered docs: a range can only be 1 dimensional in Chapel (domains introduce dimensionality by combining multiple ranges). Maybe we should specify bounded range too, so I'd suggest:

Yeah Agreed!

Aniket21mathur replied:
Made the changes :-)

Aniket21mathur commented:
> Testing and docs look good. Thanks @Aniket21mathur !

Great to see this pr merged :-)
